subxt/src/storage/storage_address.rs

```diff
@@ -54,64 +58,314 @@ pub struct Yes;
 /// via the `subxt` macro) or dynamic values via [`dynamic`].
 #[derive(Derivative)]
 #[derivative(
-    Clone(bound = "StorageKey: Clone"),
-    Debug(bound = "StorageKey: std::fmt::Debug"),
-    Eq(bound = "StorageKey: std::cmp::Eq"),
-    Ord(bound = "StorageKey: std::cmp::Ord"),
-    PartialEq(bound = "StorageKey: std::cmp::PartialEq"),
-    PartialOrd(bound = "StorageKey: std::cmp::PartialOrd")
+    Clone(bound = "Keys: Clone"),
+    Debug(bound = "Keys: std::fmt::Debug"),
+    Eq(bound = "Keys: std::cmp::Eq"),
+    Ord(bound = "Keys: std::cmp::Ord"),
+    PartialEq(bound = "Keys: std::cmp::PartialEq"),
+    PartialOrd(bound = "Keys: std::cmp::PartialOrd")
 )]
-pub struct Address<StorageKey, ReturnTy, Fetchable, Defaultable, Iterable> {
+pub struct Address<Keys: StorageMultiKey, ReturnTy, Fetchable, Defaultable, Iterable> {
     pallet_name: Cow<'static, str>,
     entry_name: Cow<'static, str>,
-    storage_entry_keys: Vec<StorageKey>,
+    keys: Keys,
     validation_hash: Option<[u8; 32]>,
     _marker: std::marker::PhantomData<(ReturnTy, Fetchable, Defaultable, Iterable)>,
 }
 
+/// A storage key, mostly used for static encoded values.
+/// The original value is only given during construction, but can be
+#[derive(Derivative)]
+#[derivative(Clone(bound = ""), Debug(bound = ""))]
+pub struct StorageKey<K: ?Sized> {
+    bytes: Static<Encoded>,
+    _marker: std::marker::PhantomData<K>,
+}
+
+impl<K: codec::Encode + ?Sized> StorageKey<K> {
+    /// Creates a new static storage key
+    pub fn new(key: &K) -> Self {
+        StorageKey {
+            bytes: Static(Encoded(key.encode())),
+            _marker: std::marker::PhantomData,
+        }
+    }
+
+    /// Returns the scale-encoded bytes that make up this key
+    pub fn bytes(&self) -> &[u8] {
+        &self.bytes.0 .0
+    }
+}
+
+/// This trait should be implemented by anything that can be used as one or multiple storage keys.
+pub trait StorageMultiKey {
```

Comment:
There's quite a lot of stuff related to this `StorageMultiKey`; I think it should all be moved to a new file and then just re-exported from `storage::address` like some other stuff is in the `mod.rs`.

Personally I think I'd just call it `StorageKey` since it's a trait representing storage keys; the Multi bit is sortof irrelevant. And then the file could be eg `storage_key.rs`

Reply:
Ah I see; there is a `StorageKey` concrete type above! I can't recall but if that's just for the static codegen then we could call it `StaticStorageKey` and then this trait can still be `StorageKey`